pg_search/src/postgres/storage/avl.rs

```diff
@@ -0,0 +1,1230 @@
+        ] {
+            assert_eq!(m.insert(k, v), Ok((None, ())));
+        }
+
+        // Delete node with two children (50). Must return the original value stored at 50.
+        assert_eq!(m.remove(&50), Some(500));
+
+        // Structure and inorder remain valid and match a reference BTreeMap after the delete.
+        let mut reference = std::collections::BTreeMap::new();
+        for &(k, v) in &[
+            (30, 300),
+            (70, 700),
+            (20, 200),
+            (40, 400),
+            (60, 600),
+            (80, 800),
+        ] {
+            reference.insert(k, v);
+        }
+        assert!(Iterator::eq(
+            m.view().iter(),
+            reference.iter().map(|(&k, &v)| (k, v))
+        ));
+
+        #[cfg(debug_assertions)]
+        m.view().assert_ok();
+    }
+
+    #[test]
+    fn test_avl_map_get_lte_basic() {
+        let mut header = AvlTreeMapHeader::default();
+        let mut buf = vec![Slot::<i32, i32>::default(); 64];
+        let mut m = AvlTreeMap::new(&mut header, &mut buf);
+
+        for &(k, v) in &[(10, 100), (20, 200), (30, 300), (40, 400)] {
+            assert_eq!(m.insert(k, v), Ok((None, ())));
+        }
+
+        // Exact hits
+        assert_eq!(m.view().get_lte(&10), Some((10, 100, ())));
+        assert_eq!(m.view().get_lte(&40), Some((40, 400, ())));
+
+        // Between keys -> greatest smaller
+        assert_eq!(m.view().get_lte(&35), Some((30, 300, ())));
+        assert_eq!(m.view().get_lte(&21), Some((20, 200, ())));
+        assert_eq!(m.view().get_lte(&11), Some((10, 100, ())));
+
+        // Below minimum
+        assert_eq!(m.view().get_lte(&-1), None);
+
+        // Above maximum
+        assert_eq!(m.view().get_lte(&99), Some((40, 400, ())));
+
+        #[cfg(debug_assertions)]
+        m.view().assert_ok();
+    }
+
+    #[test]
+    fn test_avl_map_delete_tag_preservation() {
+        let mut header = AvlTreeMapHeader::default();
+        let mut buf = vec![Slot::<i32, i32, i32>::default(); 32];
+        // Initialize the tree with a proper free list
+        let mut m = AvlTreeMap::new(&mut header, &mut buf);
+        // Set unique tags after initialization
+        for (i, slot) in m.arena.iter_mut().enumerate() {
+            slot.tag = i as i32;
+        }
+        // Insert nodes: 50 (two children), 30 (left), 60 (right, successor)
+        assert_eq!(m.insert(50, 500), Ok((None, 0)));
+        assert_eq!(m.insert(30, 300), Ok((None, 1)));
+        assert_eq!(m.insert(60, 600), Ok((None, 2)));
+        // Before delete, key=60 has tag=2
+        assert_eq!(m.view().get(&60), Some((600, 2)));
+        // Delete 50 (two children)
+        assert_eq!(m.remove(&50), Some(500));
+        // After delete, key=60 should still have tag=2, now in slot 0
+        assert_eq!(m.view().get(&60), Some((600, 2)));
+        // Slot 2 (successor) is free, but its tag remains (e.g., for reuse)
+        assert_eq!(m.arena[2].tag, 0);
+        #[cfg(debug_assertions)]
```

Comment:
I've never seen `debug_assertions` inside of a unit test before: what is the thinking there?

Reply:
That hides a call to `AvlTreeMapView::assert_ok()`, which itself is hidden behind a `#[cfg(debug_assertions)]`, so that's why it's here too.

And that function traverses the tree to make sure it has the correct structure